timings/timings.go

```diff
@@ -0,0 +1,98 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timings
+
+import (
+	"time"
+)
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
+
+// Timings represents a tree of Timing measurements for a single execution of measured activity.
+// A Timings tree object should be created at the beginning of the activity,
+// followed by starting at least one Timing measurement, and then saved at the end of the activity.
+//
+// Calling Start on the Timings objects creates a Timing and starts new
+// performance measurement. Measurement needs to be finished by calling Stop
+// function on the Timing object.
+// Nested measurements may be collected by calling Start on Timing objects. Similar
+// to the above, nested measurements need to be finished by calling Stop on them.
+//
+// Typical usagage:
+//   timingsTree := timings.New(map[string]string{"tag": "tag-value"})
+//   timing := timing.Start("computation", "...")
+//   ....
+//   nestedTiming := timing.Start("sub-computation", "...")
+//   ....
+//   nestedTiming.Stop()
+//   timing.Stop()
+//   timingsTree.Save()
+type Timings struct {
```

Comment:
Sorry - I know this is bike-shed land again but I find "Timing**s**" and "Timing" a bit too subtle. Maybe going to the suggestion of "Timings" and TimingMeasure" or "TimingsCollection" and "Timing"  (open for more ideas of course).

Reply:
I'm fine with a little bit of bikeshedding on Monday, but Timings is the one that will be referred to more often so should have the shorter or about same length name,  so not a fan of TimingsCollection

Reply:
Or maybe that's not true.  Maybe best is to land as is, and decide on first usage what are the best names.

Reply:
Spent a bit of time thinking further on this, we have actually two issues:

* a Timing is a tree of timing intervals too, that's lost with in the "Timing" name
* our Start methods don't start the object they are on, so Start/Stop is not a pair

I think:

* Timings is fine, it's bunch of timings
* s/Timing/Span/  using the 2nd def of span "the length of time for which something lasts."
* s/Start()/StartSpan() *Span/

Reply:
Thanks for thinking that through, sounds very good to me, done.